File: sdks/go/server_types.go
package togglerino

import "encoding/json"

// FlagDefinition represents a single flag as returned by the definitions endpoint.
type FlagDefinition struct {
	Key       string               `json:"key"`
	ValueType string               `json:"valueType"`
	Status    string               `json:"status"`
	Config    FlagDefinitionConfig `json:"config"`
}

// FlagDefinitionConfig holds per-environment configuration for a flag definition.
type FlagDefinitionConfig struct {
	Enabled        bool                      `json:"enabled"`
	DefaultVariant string                    `json:"defaultVariant"`
	Variants       []VariantDefinition       `json:"variants"`
	TargetingRules []TargetingRuleDefinition `json:"targetingRules"`
}

// VariantDefinition is a variant with a key and arbitrary JSON value.
type VariantDefinition struct {
	Key   string          `json:"key"`
	Value json.RawMessage `json:"value"`
}

// TargetingRuleDefinition is a targeting rule with conditions, a variant to serve,
// and an optional percentage rollout.
type TargetingRuleDefinition struct {
	Variant    string                `json:"variant"`
	Percentage *int                  `json:"percentage"`
	Conditions []ConditionDefinition `json:"conditions"`
}

// ConditionDefinition is a single condition within a targeting rule.
type ConditionDefinition struct {
	Attribute string `json:"attribute"`
	Operator  string `json:"operator"`
	Value     string `json:"value"`
}

// SegmentDefinition is a reusable segment definition.
type SegmentDefinition struct {
	Key        string                `json:"key"`
	Conditions []ConditionDefinition `json:"conditions"`
}

// definitionsResponse is the response from the definitions API endpoint.
type definitionsResponse struct {
	Flags    []FlagDefinition  `json:"flags"`
	Segments []SegmentDefinition `json:"segments"`
}

File: sdks/javascript/src/server-types.ts
/**
 * Types matching the definitions API response (camelCase JSON).
 * Used by the local evaluation engine for server-side SDKs.
 */

/** A single flag definition as returned by the definitions endpoint. */
export interface FlagDefinition {
  key: string
  valueType: string
  status: string
  defaultValue: unknown
  variants: VariantDefinition[]
  config: FlagDefinitionConfig
}

/** Per-environment configuration for a flag definition. */
export interface FlagDefinitionConfig {
  enabled: boolean
  fallthroughVariant: string
  offVariant: string
  targetingRules: TargetingRuleDefinition[]
}

/** A variant with a name and arbitrary value. */
export interface VariantDefinition {
  name: string
  value: unknown
}

/** A targeting rule with conditions, a variant to serve, and an optional percentage rollout. */
export interface TargetingRuleDefinition {
  variant: string
  percentage: number | null
  conditions: ConditionDefinition[]
}

/** A single condition within a targeting rule. */
export interface ConditionDefinition {
  attribute: string
  operator: string
  value: string
}

/** A reusable segment definition. */
export interface SegmentDefinition {
  key: string
  conditions: ConditionDefinition[]
}

/** Response from the definitions API endpoint. */
export interface DefinitionsResponse {
  flags: FlagDefinition[]
  segments: SegmentDefinition[]
}
